staging/src/k8s.io/apiserver/pkg/endpoints/request/context.go

```diff
@@ -94,3 +98,49 @@ func AuditEventFrom(ctx context.Context) *audit.Event {
 	ev, _ := ctx.Value(auditKey).(*audit.Event)
 	return ev
 }
+
+func AddEarlyAuditAnnotation(parent context.Context, key, value string) context.Context {
```

Comment:
This could be split in two functions (one for allocating the map and one for setting data).  The allocating function would need to be called in an early, distinct `WithX(handler) handler` func but it would drop the need for the setter func to return a context.

Not really sure which is better. 🤷‍♂